### PR TITLE
Onboard operator-framework/community-operators, add deploy-operator-on-openshift job

### PR DESCRIPTION
**Pull request by J0zi**
Deploy test operator to an Openshift.

### CODE REVIEW

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+base_images:
+  operator-test-playbooks:
+    name: operator-test-playbooks
+    namespace: operator_testing
```
Comment (petr-muller) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:4]: This does not seem to exist:

```console
$ oc --context api.ci --namespace operator_testing get is operator-test-playbooks --as system:admin
Error from server (NotFound): namespaces "operator_testing" not found
```

Also it does not seem to be used anywhere, so why import it as a base image?

Comment (J0zi) [reply ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:4]: Hi @petr-muller,
`make update` deleted `cluster: quay.io` line. Our image is public. Ouch, it wasn't, but it is currently, our fault. Now public. https://quay.io/repository/operator_testing/operator-test-playbooks?tab=tags

Regarding `invalid-owners-file` label. Is there something wrong. People from #forum-dp-platform told me that there is no need to be a member to create PR.

Comment (petr-muller) [reply ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:4]: It's not supposed to be a quay image, but an image present on the CI cluster. Looks like you may want to add an imagestream here: https://github.com/openshift/release/tree/master/core-services/supplemental-ci-images

`invalid-owners-file` validates the submitted `OWNERS` files. The people in the `OWNERS` must be "trusted" in order to use the features (approvals and reviews). It does not concern creating PRs.

**Comment by J0zi** [comment]
@petr-muller we successfully set an image. What is missing? Thank you.

**Comment by petr-muller** [comment]
1. You are missing a build root image (https://steps.ci.openshift.org/help/ci-operator#configuration)
2. Although you are now importing an existing image, you are not using it anywhere.

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      cpu: 100m
+      memory: 200Mi
+tag_specification:
+  name: "4.5"
+  namespace: ocp
+tests:
+- as: ansible
+  commands: test-commands.sh
+  container:
+    from: src
```
Comment (J0zi) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:20]: @petr-muller, we thought that `src` is pointer to base_image. Should we use `operator-test-playbooks` instead of `src`?

**Comment by J0zi** [comment]
> 1. You are missing a build root image (https://steps.ci.openshift.org/help/ci-operator#configuration)
> 2. Although you are now importing an existing image, you are not using it anywhere.

@petr-muller we are not going to build anything. We would like to use `ansible-runner` image and add some files there by Dockerfile. My questions are:
1. do we need to use  `build root image` anyway?
2. Where should be Dockerfile located? In our original repo?

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml (hunk 2 above)
Comment (petr-muller) [reply ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:20]: I don't know what `operator-test-playbooks` is, so I have no idea, You'd probably want to use the image that has `test-commands.sh` (if that lives in `operator-framework/community-operators` then `src` or something derived from it is likely a good candidate) together with all dependencies (usually supplied in the build root).

**Comment by petr-muller** [comment]
> @petr-muller we are not going to build anything. We would like to use `ansible-runner` image and add some files there by Dockerfile.

"like to use" use for what? You mean to ultimately execute the `test-commands.sh` script from your repository? If that is the case, then I think you need to build your own image (https://steps.ci.openshift.org/help/ci-operator#buildroot) from `ansible-runner` and use *that* as a build root. That means that the sources of your repository will be cloned to that image, the resulting image will be called `src`, and you an use it as a image for the test.

 My questions are:
> 
>     1. do we need to use  `build root image` anyway?

Yes.

>     2. Where should be Dockerfile located? In our original repo?

Yes.

**Comment by J0zi** [comment]
@petr-muller, reharse saying `COPY test-commands.sh /test-commands.sh`
but the [Dockerfile](https://github.com/operator-framework/community-operators/blob/master/scripts/ci/Dockerfile.ci-operator) is updated already, so prow is using old cached one. I tried to restart job and reopen a PR. Still not using actual Dockerfile.

Another issue is with ordered-prow-config, jobs are generated by `make update`, I have no impact on formatting.

**Comment by petr-muller** [comment]
/test pj-rehearse

Interesting, looks like a corner case in input resolution where we don't detect that the target repo HEAD has moved and ci-operator will reuse previous artifacts because it thinks the inputs did not change. I'll make an issue for it.

I deleted the [temporary namespace](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308739287130312704#1:build-log.txt%3A5) (that acts like a cache):

```console
$ oc delete namespace ci-op-yhz8xfxv --as system:admin
namespace "ci-op-yhz8xfxv" deleted
```

As an author of the PR, ci-operator gives you admin access to that namespace so that you can delete it yourself, too.

For formatting checks, please rebase. The checks are driven by configuration files in the repository and run as if your PR would be merged. You probably have config files stale and run `make update` over that.

**Comment by J0zi** [comment]
> I deleted the [temporary namespace](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308739287130312704#1:build-log.txt%3A5) (that acts like a cache):
> 
> ```
> $ oc delete namespace ci-op-yhz8xfxv --as system:admin
> namespace "ci-op-yhz8xfxv" deleted
> ```
> 
> As an author of the PR, ci-operator gives you admin access to that namespace so that you can delete it yourself, too.

How can I connect to this openshift? Is there some virtual console or some config/credentials I can use on my laptop? O some readme?

**Comment by petr-muller** [comment]
The job always links to the full namespace URL https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308759353351737344#1:build-log.txt%3A5

All `openshift` org members have access to these clusters after doing a GH OAuth round.

**Comment by J0zi** [comment]
> All `openshift` org members have access to these clusters after doing a GH OAuth round.

@petr-muller, could you please be more specific? Let's assume that they approved my membership to `openshift`. I will see some "tab" and get virtual console? Where?

Petr, could you please have a look why `test-commands.sh: command not found`. It was successfully copied with Dockerfile. Should be present.
Thanks a lot.

**Comment by petr-muller** [comment]
> @petr-muller, could you please be more specific? Let's assume that they approved my membership to `openshift`. I will see some "tab" and get virtual console? Where?

I don't know where you would expect to see any tab, and I don't know what you mean by "virtual console", sorry.

If you are an `openshift` member, then:

1. You follow the link from the [log](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308835495765086208#1:build-log.txt%3A5), e.g. https://console-openshift-console.apps.build01.ci.devcluster.openshift.com/k8s/cluster/projects/ci-op-ddx5fbzy
2. This link leads to an openshift console of the given cluster. First time you access the cluster, you are given an option to login via a GitHub OAuth procedure. This makes you an account on the cluster and you can interact with it via the web console or `oc`.
3. As an author of the PR, ci-operator [creates RBACs](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308835495765086208#1:build-log.txt%3A8) to give you admin in the given temporary namespace.
4. This is useful for inspecting the built images or other resources like pods.

> Petr, could you please have a [look](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-ansible/1308759353351737344/build-log.txt) why `test-commands.sh: command not found`. It was successfully copied with Dockerfile. Should be present.

This is becoming frustrating. I have explained how this works several times in this PR, and I have linked the documentation that explains everything else. Please read the documentation.

1. There is no reason to copy `test-commands.sh` into the build root. The `src` image will contain the cloned working copy of your repository by default ([doc](https://steps.ci.openshift.org/help/ci-operator#artifacts)). 

```console
$ podman run --entrypoint /bin/ls registry.build01.ci.openshift.org/ci-op-ddx5fbzy/pipeline:src -- scripts/ci/test-commands.sh
scripts/ci/test-commands.sh
```

2. The working dir of the pipeline images is the root of the cloned repository ([doc](https://steps.ci.openshift.org/help/ci-operator#tests)). Of course the file is where you copied it:

```console
$ podman run --entrypoint /bin/pwd registry.build01.ci.openshift.org/ci-op-ddx5fbzy/pipeline:src -- 
/go/src/github.com/operator-framework/community-operators

$ podman run --entrypoint /bin/ls registry.build01.ci.openshift.org/ci-op-ddx5fbzy/pipeline:src -- /test-commands.sh
/test-commands.sh
```

**Comment by J0zi** [comment]
@petr-muller We would like to run podman pull/push in container. Soemthing like that:
```
podman run --privileged --rm -it quay.io/operator_testing/operator-test-playbooks podman --storage-driver=vfs pull docker.io/centos
ERRO[0000] unable to write system event: "write unixgram @0002e->/run/systemd/journal/socket: sendmsg: no such file or directory" 
Trying to pull docker.io/centos...
Getting image source signatures
Copying blob 3c72a8ed6814 done
Copying config 0d120b6cca done
Writing manifest to image destination
Storing signatures
0d120b6ccaa8c5e149176798b3501d4dd1885f961922497cd0abef155c869566
```

We tried something like that
```
securityContext:
  runAsUser: 0
  privileged: true
```
But it is removed by `make update`
can you give us a hint?

**Comment by petr-muller** [comment]
We currently do not support privileged containers nor we plan to add such support.

**Comment by J0zi** [comment]
@petr-muller, we pushed a new image and restarted tests. But tests are using an old image, so I tried to delete namespace according your advice, but have no rights. Am I missing something? Thank you in advance:)

```
Logged into "https://api.build01.ci.devcluster.openshift.com:6443" as "J0zi" using the token provided.

You have access to the following projects and can switch between them with 'oc project <projectname>':

    ci-op-5wn8p8xn
    ci-op-89zmnvyb
    ci-op-i7gmndvt
    ci-op-jvz54l8k
  * ci-op-mvpxhslc
    ci-op-vpcslbtt

Using project "ci-op-mvpxhslc".
 oc project ci-op-5wn8p8xn                                                                                                 
Now using project "ci-op-5wn8p8xn" on server "https://api.build01.ci.devcluster.openshift.com:6443".
 oc delete namespace 5wn8p8xn                                                                                              
Error from server (Forbidden): namespaces "5wn8p8xn" is forbidden: User "J0zi" cannot delete resource "namespaces" in API group "" in the namespace "5wn8p8xn"
```

**Comment by petr-muller** [comment]
The RBAC is there, having admin maybe does not allow deleting a namespace? Not sure. Can you delete individual builds?

``` console
$ oc  --context build01 --namespace ci-op-i7gmndvt describe rolebinding ci-op-author-access
Name:         ci-op-author-access
Labels:       <none>
Annotations:  <none>
Role:
  Kind:  ClusterRole
  Name:  admin
Subjects:
  Kind  Name  Namespace
  ----  ----  ---------
  User  J0zi
```

**Comment by petr-muller** [comment]
Can you be more specific about "But tests are using an old image". What's the symptom? Old image of what?

**Comment by J0zi** [comment]
@petr-muller, regarding old image: It is about our image, which I edited and pushed. We are using it via [Dockerfile](https://github.com/openshift/release/pull/12072/files#diff-fc2344fa1d1705ebe86961663cb21874R8) and basically it is `FROM quay.io/operator_testing/operator-test-playbooks`.
So after image was rebuilt on my local machine and pushed, I see pods are using the old one. So wondering how to pull the actual version of our `operator-test-playbooks` image during pj-rehearse test.

**Comment by petr-muller** [comment]
Yes, the root image would be cached if the only thing changed would be its base image. You should be able to delete the builds in the temporary namespace:

```console
$ oc  --context build01 --namespace ci-op-i7gmndvt delete build root src
```

**Comment by J0zi** [comment]
@petr-muller I would like to ask. How is it possible that `src-test container setup` is green [here](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-src-test/1313018754262634496) despite the fact that `ansible-pull` command failed 6 lines before end of [this](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-src-test/1313018754262634496/artifacts/src-test/container-logs/test.log) log?

**Comment by petr-muller** [comment]
> @petr-muller I would like to ask. How is it possible that `src-test container setup` is green in the task `ci/rehearse/operator-framework/community-operators/master/src-test` [here](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-src-test/1313018754262634496) despite the fact that `ansible-pull` command failed 6 lines before end of [this](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/12072/rehearse-12072-pull-ci-operator-framework-community-operators-master-src-test/1313018754262634496/artifacts/src-test/container-logs/test.log) log? How can we make it fail in case of failed command?

How am I supposed to know? It is *your* script. We just take the exit code of whatever you run.

**Comment by J0zi** [comment]
@petr-muller, please approve and let it merge. If anything is needed, let me know. Thank you for your help during the PR, we appreciate it.

**Comment by J0zi** [comment]
@bbguimaraes @petr-muller, please merge if you find some time. Thanks a lot. You will help us to meet deadlines on official bundle support release date for 2 streams - community operators and upstream community operators.

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  namespace: ocp
+tests:
+- as: deploy-operator-on-openshift
+  commands: |
+    echo "Need to clone default branch, cloning..."
+    mkdir -p /tmp/oper
+    cd /tmp/oper
+    git clone https://github.com/J0zi/community-operators.git
+    cd community-operators
+    git checkout RHO-716-release-on-openshift
```
Comment (petr-muller) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:25]: Why is this needed? This is an anti-pattern that we discourage. Why does this need to depend on out-of-tree fork?

Comment (J0zi) [reply ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:25]: @petr-muller the test is in beta and we suppose to update it frequently until we get stable version. Then we definitely remove this part, merge master and use it. But until then, I will craft my branch until satisfied. Let me know if it is possible until next PR.

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      cpu: 1000m
+      memory: 4Gi
+tag_specification:
+  name: "4.6"
+  namespace: ocp
+tests:
+- as: deploy-operator-on-openshift
+  commands: |
+    ls
+    /go/src/github.com/operator-framework/community-operators/scripts/ci/openshift-deploy.sh
```
Comment (petr-muller) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:21]: Please use the relative path (workdir will be the a working copy of the repo, but we do not guarantee the exact location)

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml (hunk 3 above)
Comment (petr-muller) [reply ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml]: It's possible, but like you spotted, you will likely encounter artifact reuse if the only thing that changed between individual test runs is the out-of-tree code.

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  project_image:
+    dockerfile_path: scripts/ci/Dockerfile.ci-operator
+resources:
+  '*':
+    requests:
+      cpu: 1000m
+      memory: 4Gi
+tag_specification:
+  name: "4.6"
+  namespace: ocp
```
Comment (petr-muller) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:16]: What are your expectations about the installed cluster versions and stability? The `tag_specification` stanza means that `ci-operator` will assemble an ephemeral release (in this case, 4.6) from the content of the OCP integration imagestream (basically, this corresponds to HEADs of all 4.6 branches). These releases may be more unstable/unreliable than the releases vetted by the release controller, or even an already-released-release (what OCP customers have). If your goal is to provide stable feedback to operator authors and not wear them down with running on OCP integration releases, you should consider using a ` releases` stanza (https://steps.ci.openshift.org/help/ci-operator#release) that allows you to specify what kind of "existing" release you would like to use. It seems to me that a nightly, or even a released image should be a better choice than running on integration imagestream OCP.

File: ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+base_images:
+  cli:
+    name: "4.6"
+    namespace: ocp
+    tag: cli
+build_root:
+  project_image:
+    dockerfile_path: scripts/ci/Dockerfile.ci-operator
+releases:
+  previous:
```
Comment (petr-muller) [inline ci-operator/config/operator-framework/community-operators/operator-framework-community-operators-master.yaml:10]: Call this `latest`

> Releases may be named, with two names holding special meaning. In ordinary end-to-end tests, the latest release describes the version that will be installed before tests are run.

**Comment by petr-muller** [comment]
Also note that `do-not-merge/invalid-owners-file` label is a merge blocker - either remove that person from `OWNERS`, or make them enroll.

**Comment by J0zi** [comment]
@petr-muller we stick with ephemeral image for now. Time is running out. Thank for the hint, we will tweak it in a next PR. Can you help us merge it?

**Comment by petr-muller** [comment]
> @petr-muller we stick with ephemeral image for now. Time is running out. Thank for the hint, we will tweak it in a next PR. Can you help us merge it?

There's no good reason for sticking with ephemereal images. This use case is exactly what the `releases:` stanza is for.

**Comment by petr-muller** [comment]
/label tide/merge-method-squash
/retitle Onboard operator-framework/community-operators, add deploy-operator-on-openshift job

**Review by petr-muller — APPROVED** [review]
LGTM
